Support gpg key expiration verification

lukpueh opened the pull request:
**Fixes issue #**:
Takes over #245  (Thanks @michizhou, for prototyping this!)
Fixes #236 

**Description of the changes being introduced by the pull request**:
This PR adds support for parsing creation date and validity period of gpg primary and sub public keys from successfully verified self-signature subpackets. Two new fields are added to gpg public keys:
- `creation_time`, and
- `validity_period`

If the gpg packets carry ambiguous information we resolve as follows:
- For subpackets we only use the last subpacket in the list and prioritize hashed over unhashed subpackets (see 82368677c8abe98adfa3cb97fcd71c4957e82a95)

- For User IDs and certificates we take the key expiration time from the most recent certificate, i.e. the certificate with the highest signature creation time. Additionally, we prioritize certificates with primary user id flag set True. If the ultimately prioritized certificate does not have a key expiration time subpacket, we don't assign one, even if there were certificates of lower priority carrying that subpacket (see 5f4dfbda6ae2e865f1c8d4b5d8b6eaff6fbea425).

The PR also modifies the gpg API functions `gpg_verify_signature`, to check and raise a custom `KeyExpirationError` if the passed verification key has expired, and `gpg_sign_object` to raise a custom `CommandError` if the executed gpg command exits with non-zero, e.g. because the signing key has expired (see ae047a39fad66400e1e0d705c4e53228226de3cb and 18218239c972428864026a95bef15c9070e6fbb2).

The docstrings of calling functions are adopted accordingly. And in verifylib, where we don't count links with invalid signatures towards the threshold, we add `KeyExpirationError` as additional reason to skip a link (see 19eb3c4b6f0c014287456f046cff95e12577f174).

*OT: 23cb4d7e62233b5891e2d7c5d5b6bc07fb1d909d fixes bugs in the test aggregator script and re-adds some test verbosity that was removed in earlier commits, i.e. it prints which gpg version is available on the test system). We don't have to merge it with this PR (but we might as well).*

See commit messages for more details.




**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

michizhou committed:
ENH: gpg:common: parse and verify expiration dates

The current implementation doesn't parse the expiration dates of keys in
the keyring. Add code to parse and verify that keys used within in-toto
are not expired.

michizhou committed:
Added signature information extraction for public keys

Enhanced parsing of public key bundles with additional information parsing of
key expiration times from signatures to provide master public keys and subkeys
with information needed to ensure that no expired keys are used within in-toto.
Also added tests to ensure proper key expiration time parsing from signatures
and new public key with a set Primary User ID flag for tests.

Commit:
Remove trailing whitespace

Commit:
Remove obsolete gpg data

Commit:
Remove unused imports in gpg subpackage and test

Commit:
Remove expiration parsing in parse_pubkey_payload

Commit:
Remove obsolete SIGNATURE_TYPE_PARSING constant

Commit:
Rename gpg pubkey creation and expiration fields

creation_date --> creation_time (it's a unix timestamp after
all)
expiration --> validity_period (validity period seems less
ambiguous, it is the the number of seconds after the key creation
time that the key expires)

Commit:
Remove expiration parsing in get_pubkey_payload

Commit:
Remove gpg signature key_expire_time field

Remove redundant parsing of key expiration time subpacket in
gpgp.common.parse_signature_packet. This and other subpackets are
also added to  the signature's opaque "info" field. Which can
be used for further intermediary processing of signature data
that does not need to end up in the ultimate signature.

This commit also changes the "info"'s "subpackets" list to a
dictionary, to consider the last subpacket of a type only, in case
there are more. Hashed subpackets are prioritized over unhashed.
This is the same subpacket ambiguity resolution scheme we use for
issuer keyid and fingerpint.

Commit:
Refactor gpg certified key information assignment

We implement the following ambiguity resolution scheme, in case there
are multiple self-certificates with conflicting key information:

We take the key expiration time from the most recent certificate,
i.e.  the certificate with the highest signature creation time.
Additionally, we prioritize certificates with primary user id flag
set True. Note that, if the ultimately prioritized certificate does
not have a key expiration time subpacket, we don't assign one, even
if there were certificates of lower priority carrying that
subpacket.

Commit:
Add test keys with key expiration subpackets

Primary keyid is:
C242A830DAAF1C2BEF604A9EF033A3A3E267B3B1

It has 3 User IDs that set the expiration date for the primary
key:
:user ID packet: "Test Expiration I <test@expir.one>"
:user ID packet: "Test Expiration II <test@expir.two>" (primary)
:user ID packet: "Test Expiration III <test@expir.three>"

And two subkeys, one with expiration date and one without.

Commit:
Add gpg expiration date parsing tests

Test assignment of primary key and subkey expiration dates,
including ambiguity resolution.
See docstring of `test_assign_certified_key_info_expiration` for
testing limitations.

Commit:
Replace default-key with local-user in gpg sign

Pass keyids to gpg signing command using the --local-user
option instead of the --default-key option for consistent
handling of expired keys across gpg versions.

Testing the following command
```
echo bla | gpg --homedir . --detach-sign --default-key 2C199C7C
```
on "gpg (GnuPG/MacGPG2) 2.2.10" and "gpg (GnuPG) 1.4.18", only the
latter fails when trying to sign with an expired key, with the
following message on stderr:
  gpg: no default secret key: unusable secret key
  gpg: signing failed: unusable secret key

The newer version ignores the passed keyid and sign with the
keyring's default keyid, with the following message on stderr:
  gpg: Warning: not using '2C199C7C' as default key: No secret key
  gpg: all values passed to '--default-key' ignored

With --local-user signing fails on both versions.

Commit:
Remove obsolete gpg key expiration tests

Commit:
Make gpg KeyExpirationError message timezone aware

As per RFC4880 section 3.5 time fields, such as the key creation
time contain the number of seconds elapsed since midnight, 1
January 1970 UTC.

This commit modifies the error message in a KeyExpirationError
to use timezone UTC for creation and calculated expiration date
and adds the timezone specifier to the  date trings.

The timezone name, i.e. "UTC" is hardcoded, because the %Z
specifier uses different names on different python versions.

Expiration assessment is unaffected, as it uses time.time(), which
is compatible with creation time (i.e. seconds since epoch).

Commit:
Adopt new gpg errors in caller docstrings

Recent changes modified gpg.functions.gpg_verify_signature and
gpg.function.gpg_sign_object to raise new error types
- in_toto.gpg.exceptions.CommandError
- in_toto.gpg.exceptions.KeyExpirationError

This commit adopts the docstrings of calling functions to mention
that these error types might be raised.

Commit:
Adopt expiration in verify_link_signature_thresholds

Adopt verifylib.verify_link_signature_thresholds to skip (and
log) links files that were signed with an expired key.

Adds test.

Commit:
Re-add gpg verbosity in runtests check_usable_gpg

844366ac5cf41ef7665f4df229ac4616d092a9de and
0a63a0fa1191033cce5f97a1062df1e052f1394a modified the test
aggregators gpg availablity check to send the output to devnull,
with the goal to remove noise in the testlog.

However, knowing the available version is also helpful for
debugging issues on e.g. CI systems.

The function now also checks for gpg2 and only sets TEST_SKIP_GPG
if neither gpg2 nor gpg are available.

This commit also updates the caught error to type OSError,
which encompasses the possible errors cross-platform/version.
- OSError (on Python2)
- FileNotFoundError (on Python 3)
- WindowsError (on windows)

Commit:
Fix _get_verified subkeys docstring and test

The function returns a dict of pubkeys, not a list or an ordered
dict.

Commit:
Raise custom CommandError in gpg_sign_object

Raise custom CommandError in gpg_sign_object if gpg command
returns a non-zero return value and include command's stderr
in exception message.

Test with expired test key.

The commit also removes earlier added key export and expiration
date checks, this was unnecessary (gpg can't sign with expired
keys anyway), added redundancy (the same key was exported twice),
issued a warning that was not relevant to signing.

Commit:
Raise custom KeyExpirationError in gpg_verify_signature

Raise custom KeyExpirationError in
gpg.functions.gpg_verify_signature if the passed key has a
creation time and validity period field and the two added
values are smaller then the current seconds since epoch, i.e.
the key has expired.

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -334,11 +337,61 @@ def _assign_certified_key_info(bundle):
+
+      # No signature creation time, no way to resolve potential ambiguities,
+      # go to next certificate
+      # TODO: As per the spec this cannot not happen, should we blow up?
```
I do think we'd like to blow up in this case, or log a warning at least.

lukpueh replied:
Good point! I slightly refactored the parsing in https://github.com/in-toto/in-toto/pull/266/commits/f071c95f954a9beb7398470f4a4391b11d61b8f8. Please see the commit message for more details.

We still don't blow up, but we log the problem, and with the refactor, blowing up will be easier  in the future  (also see the updated `TODO` comment).

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -334,11 +337,61 @@ def _assign_certified_key_info(bundle):
+      # that, if the ultimately prioritized certificate does not have a key
+      # expiration time subpacket, we don't assign one, even if there were
+      # certificates of lower priority carrying that subpacket.
+      tmp_validity_period = \
```
Is there a reason to rename `key_expiration` to `validity_period`? I think it may be confusing. What do you think?

lukpueh replied:
As per the [RFC 4880](https://tools.ietf.org/html/rfc4880#section-5.2.3.6):
> **5.2.3.6.  Key Expiration Time**
> The validity period of the key.  This is the number of seconds after the key creation time that the key expires. [...]

So the spec calls it both. But IMO "expiration" or "expiration time" is more ambiguous. It could easily be both, the number of seconds between creation and expiration (which it is), or a set date on which it expires (which it isn't). I think `validity_period` is less ambiguous in that regard. 

That said, I'm not overly passionate about it. :)

SantiagoTorres replied:
Hmm, I see what you mean. I just found myself jumping where `tmp_validity_period` was defined and where it was checked, as I would read on the dict the `KEY_EXPIRATION_SUBPACKET`.

I know the RFC is not the clearest piece of text ever written (just wait for v5 keys!) but yeah...

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -347,12 +400,10 @@ def _get_verified_subkeys(bundle):
-    TODO: Extract additional information about subkeys from the signatures
-    (see inline TODO below).
+    the passed bundle in order to enrich subkeys with additional information
+    (e.g. expiration dates). Only valid (i.e. parsable) subkeys that are
```
Note to self (and everybody) `parsable` is a word (unlike parse-able, which is an ugly composite word)

SantiagoTorres commented on in_toto/gpg/common.py:
```diff
@@ -660,6 +718,7 @@ def parse_signature_packet(data, supported_signature_types=None,
   # conflict resolution scheme that makes more sense.
   # (see RFC4880 5.2.4.1.)
   # Below we only consider the last and favor hashed over unhashed subpackets
+  # TODO: Should we warn if a we use an unhashed subpacket?
```
I think this would be a nice-to-have if possible.

lukpueh replied:
Yes, it would be. But I'm not sure where to best place the warning. Would it be okay to just keep the `TODO` note for now and revisit at a later time?

SantiagoTorres reviewed (approved):
I like this!

one small concern that I have is that this seems to have subsumed #265. Has it?

lukpueh commented:
Re #265: Yes, partially. It fixes a NameError and a cross-platform bug. See conversation in #264. 

But the fix for those particular bugs as proposed in #265 is actually incorrect, however, #265 provides some other modifications that should not be discarded (I volunteer to review #265 once this PR here is merged).

Btw., to make the mess complete, I also touched the mentioned lines in https://github.com/in-toto/in-toto/pull/279/commits/12d2f23745b39b8dcb3c4203c38355223fe2ce13 and https://github.com/in-toto/in-toto/pull/279/commits/c666aeb0bc4ae1014835979bad5119936792f21c of #279. But don't worry, I'll clean up accordingly, depending on which PR gets merged first. ;)

Commit:
Make gpg signature creation subpacket mandatory

Following a review comment by @SantiagoTorres, this commit modifies
in_toto.gpg.common to explicitly parse the signature creation time
subpacket in the signature parsing function, instead of
`_assign_certified_key_info` where it is used, and raise
ValueError if it is not present, instead of just ignoring it.

As a consequence `_assign_certified_key_info` does not have to
test and case handle if the subpacket is not there. However,
`_assign_certified_key_info` will still skip signatures with
missing creation time subpacket, because it skips
all un-parseable signatures.

This commit also updates a TODO comment that refers to an upcoming
revision of the exception taxonomy.
